### ISSUE TITLE
map reduce doesnt support start_key + end_key aliases

### ISSUE THREAD

**Issue by daleharvey** [pouchdb/pouchdb#4154]

**Comment by nolanlawson** [comment]
fixed by https://github.com/pouchdb/pouchdb/commit/f8629e3be86e619411598848b368ff9cd7846e56

**Comment by daleharvey** [comment]
That was for allDocs, do we get this for map reduce for free via allDocs? I know m/r is based on allDocs but still figured there would be some extra work

**Comment by nolanlawson** [comment]
Oh my bad, no we don't. In fact, this is another good "first timer" bug, because it is probably a one-line code change to pass in the parameter.

**Comment by daleharvey** [comment]
Going to reserve the help wanted tag for when issues are written in in a clear way for new contributors

**Comment by nolanlawson** [comment]
Steps for first timers
---

I've created [a branch](https://github.com/pouchdb/pouchdb/tree/4154) with a unit test that will fail until this issue is fixed. Basically what we want is for `start_key` and `end_key` to be synonyms of `startkey` and `endkey` for the `query()` API, but they're currently not.

To get started:

1. Make sure you install Node, npm, and git. Due to the leveldown dependency, you may need to install some additional packages, but the installer will tell you which one you're missing (see [this page](https://github.com/nodejs/node-gyp#installation) for details if you get stuck).
2. Run `git clone git@github.com:pouchdb/pouchdb.git` to clone pouchdb, then `cd pouchdb` and `git checkout 4154` to check out the branch called `4154`.
3. To run the entire mapreduce test suite, run `TYPE=mapreduce npm test`. To run just this single test, run `TYPE=mapreduce GREP=4154 npm test`. You can also run `npm run dev` and it will provide instructions for running the exact same tests in your browser.
4. You should notice that the test is failing. I see: `AssertionError: Startkey is inclusive: expected [ Array(5) ] to have a length of 4 but got 5`.
5. Next you will need to add the synonyms. As it turns out, `startkey` and `endkey` exist in the `allDocs()` API as well, and `query()` eventually calls `allDocs()` (because we implemented map/reduce over `allDocs()`), so you just need to find where mapreduce calls `allDocs()` and ensure that `start_key` is treated the same as `startkey` and `end_key` is treated the same as `endkey`. 

More tips:

I recommend starting to look in `lib/mapreduce/index.js` and look for code that calls a remote CouchDB over HTTP and passes in a bunch of options, including `startkey` and `endkey`. It should be easiest to get the test passing for CouchDB first, because you just need to pass in the extra options. 

Then to get it passing for PouchDB, search the code for where user-supplied `opts` containing `startkey`/`endkey` are translated into another options object that is eventually passed into `allDocs()`. (Note: you may not want to actually pass the synonyms themselves into `allDocs()`; it might be easier to just massage them before you pass them in. You'll see when you get there!)

Feel free to play around a bit - if you're testing in the browser, you can click on the "play" button to the right of a particular test, and keep refreshing every time you change something (it will auto-reload). Good luck! :smiley:

**Comment by nickcolley** [comment]
I've updated our tags from 'help wanted' to 'first timers only' to better show who these issues are for.

**Comment by nolanlawson** [comment]
Ah okay!

**Comment by nicolasbrugneaux** [comment]
I was trying to tackle this issue :) But turns out I can't get the test running, it's stuck at 
```
❯ TYPE=mapreduce npm test

> pouchdb@4.0.1-prerelease test /Users/nicolasb/workspace/pouchdb
> npm run version && npm run jshint && bash bin/run-test.sh


> pouchdb@4.0.1-prerelease version /Users/nicolasb/workspace/pouchdb
> node ./bin/get-version.js


> pouchdb@4.0.1-prerelease jshint /Users/nicolasb/workspace/pouchdb
> jshint -c .jshintrc bin/ lib/ tests/

Waiting for host to start .................................................
```

**Comment by noopkat** [comment]
nice issue setup, peeps! 🙌

**Comment by nickcolley** [comment]
Fixed by @nicolasbrugneaux with d977e64 :tada:
